### PR TITLE
fix(ext/http): resource leak if request body is not consumed

### LINKED ISSUE #11926: hyper server leaks RequestResource objects
Can be observed by logging the output of `Deno.core.resources()` after every request.

Good:
```
{"0":"tcpListener","2":"httpConnection","3":"responseSender"}
{"0":"tcpListener","5":"httpConnection","6":"responseSender"}
{"0":"tcpListener","8":"httpConnection","9":"responseSender"}
{"0":"tcpListener","11":"httpConnection","12":"responseSender"}
```
Bad:
```
{"0":"tcpListener","2":"httpConnection","4":"responseSender","3":"request"}
{"0":"tcpListener","6":"httpConnection","8":"responseSender","3":"request","7":"request"}
{"0":"tcpListener","10":"httpConnection","12":"responseSender","3":"request","7":"request","11":"request"}
{"0":"tcpListener","14":"httpConnection","16":"responseSender","3":"request","7":"request","11":"request","15":"request"}
```
It's more obvious with http/2 but http/1 POST + request body shows the same behavior.

(http/1 GET does not but I suspect it takes a shortcut and doesn't create a `RequestResource`.)

I'm unclear if the `RequestResource` objects are cleaned up when the connection object is closed, but they're definitely not being cleaned up after the response is sent.

### PR DESCRIPTION
**Pull request by bartlomieju**
Fixes #11926

### CODE REVIEW

**Comment by bartlomieju** [comment]
On a related note: there are two code paths in `createRespondWith` that can throw `TypeError` - I believe similar cleanup should be done in those paths. @lucacasonato what do you think?

File: cli/tests/unit/http_test.ts [hunk 1]
```diff
@@ -882,3 +882,45 @@ unitTest(
...
+      body: "hello world",
+    });
+    await response.text();
+    const resourcesAfter = Deno.resources();
+    // verify that the only new resource is "httpConnection", to make
+    // sure "request" resource is closed even if its body was not read
+    // by server handler
+
+    for (const rid of Object.keys(resourcesBefore)) {
+      delete resourcesAfter[Number(rid)];
```
Comment (bnoordhuis) [inline cli/tests/unit/http_test.ts:917]: Why Number? `resourcesAfter` is not an array, right?

File: ext/http/01_http.js [hunk 2]
```diff
@@ -297,6 +297,13 @@
 
           ws[_eventLoop]();
         }
+      } else {
+        // Try to close "request" resource. It might have been already consumed,
+        // but if it hasn't been we need to close it here to avoid resource leak.
+        try {
+          SetPrototypeDelete(httpConn.managedResources, requestRid);
+          core.close(requestRid);
+        } catch { /* pass */ }
```
Comment (bnoordhuis) [inline ext/http/01_http.js:306]: I don't really understand why it's inside a swallow-everything try/catch - seems like that could hide genuine bugs - but it's a pattern that's used elsewhere in this file so :shrug:

**Review by bnoordhuis — APPROVED** [review]

File: cli/tests/unit/http_test.ts (hunk 1 above)
Comment (bartlomieju) [reply cli/tests/unit/http_test.ts:917]: TS was complaining about `rid` being any type 🤷

File: ext/http/01_http.js (hunk 2 above)
Comment (bartlomieju) [reply ext/http/01_http.js:306]: `core.close()` should only throw `BadResource` when the thing has already been closed.

**Pull request merged**